R/rows.R

```
@@ -0,0 +1,277 @@
+#'   The default is `FALSE` with an informative message.
+#'
+#' @return A tbl object of the same structure as `x`.
+#'   On mutable backends, same as `x` if `inplace = TRUE`.
+#'
+#' @name rows
+#' @examples
+#' data <- tibble(a = 1:3, b = letters[c(1:2, NA)], c = 0.5 + 0:2)
+#' data
+#'
+#' # Insert
+#' rows_insert(data, tibble(a = 4, b = "z"))
+#' try(rows_insert(data, tibble(a = 3, b = "z")))
+#'
+#' # Update
+#' rows_update(data, tibble(a = 2:3, b = "z"))
+#' rows_update(data, tibble(b = "z", a = 2:3), by = "a")
+#'
+#' # Variants: patch and upsert
+#' rows_patch(data, tibble(a = 2:3, b = "z"))
+#' rows_upsert(data, tibble(a = 2:4, b = "z"))
+#'
+#' # Delete and truncate
+#' rows_delete(data, tibble(a = 2:4))
+#' rows_delete(data, tibble(a = 2:4, b = "b"))
+#' rows_delete(data, tibble(a = 2:4, b = "b"), by = c("a", "b"))
+#' rows_truncate(data)
+NULL
+
+
+#' rows_insert
+#'
+#' `rows_insert()` adds new rows.
+#' This operation corresponds to `INSERT` in SQL.
+#' If `by` is non-empty, no two rows with the same values in the key columns
+#' are permitted.
+#' @rdname rows
+#' @export
+rows_insert <- function(x, y, by = NULL, ..., copy = FALSE, inplace = NULL) {
+  check_rows_args(x, y, by)
+  UseMethod("rows_insert")
+}
+
+#' @export
+rows_insert.data.frame <- function(x, y, by = NULL, ..., copy = FALSE, inplace = NULL) {
+  y <- auto_copy(x, y, copy = copy)
+  y_key <- df_key(y, by)
+  x_key <- df_key(x, names(y_key))
+  df_inplace(inplace)
+
+  if (has_length(y_key)) {
+    idx <- vctrs::vec_match(y[y_key], x[x_key])
+    bad <- which(!is.na(idx))
+    if (has_length(bad)) {
+      abort(class = "dplyr_rows_insert_duplicate",
+        "Attempting to insert duplicate rows.",
+        location = bad
+      )
+    }
+  }
+
+  vctrs::vec_rbind(x, y)
+}
+
+#' rows_update
+#'
+#' `rows_update()` updates existing rows.
+#' This operation corresponds to `UPDATE` in SQL.
+#' `by` is mandatory and defaults to the first column in `y`.
```

I find this statement (here and all the other instances) confusing because it seems to contradict itself. If `by` is mandatory, why is the signature `by = NULL` and we silently default to the first column of `y`?

Related to my proposal to create a small exported helper for the purposes of making this default transparent.

---

Maybe:

```suggestion
#' `by` is mandatory, if omitted the first column in `y` is used.
```

---

That's definitely better. I think it probably needs to message to help people remember that the default behaviour for `by` for `row_*()` functions differs from that of the `*_join()` functions.